File: 3/3-1/Java/CRC.java
import java.util.Arrays;

public class CRC {
	private String gxStr = "10001000000100001";

	public String GetRemainderStr(String dividendStr, String divisorStr) {
		int dividendLen = dividendStr.length();
		int divisorLen = divisorStr.length();
		for (int i = 0; i < divisorLen - 1; i++) {
			dividendStr += "0";
		}
		char str1[] = dividendStr.toCharArray();
		char str2[] = divisorStr.toCharArray();
		for (int i = 0; i < dividendLen; i++) {
			if (str1[i] == '1') {
				str1[i] = '0';
				for (int j = 1; j < divisorLen; j++) {
					if (str1[i + j] == str2[j]) {
						str1[i + j] = '0';
					} else {
						str1[i + j] = '1';
					}
				}
			}
		}
		String remainderStr = Arrays.toString(str1).replaceAll(", ", "").substring(dividendLen + 1,
				dividendLen + divisorLen);
		return remainderStr;
	}

	public String Send() {
		String dataStr = "01100000000000111110000000000110";
		System.out.println("待发送的数据信息二进制比特串为：" + dataStr);
		System.out.println("CRC-CCITT对应的二进制比特串为：" + gxStr);

		String remainderStr = GetRemainderStr(dataStr, gxStr);

		String crcStr = remainderStr;
		String sendFrameStr = dataStr + remainderStr;
		
		System.out.println("生成的CRC-Code为: " + crcStr);
		System.out.println("带校验和的发送帧为: " + sendFrameStr);
		System.out.println();
		return sendFrameStr;
	}

	public void Receive(String sendFrameStr) {
		int sendFrameLen = sendFrameStr.length();
		int gxLen = gxStr.length();

		String dataStr = sendFrameStr.substring(0, sendFrameLen - gxLen + 1);
		String crcStr = sendFrameStr.substring(sendFrameLen - gxLen + 1);
		System.out.println("接收的数据信息二进制比特串为：" + dataStr);
		System.out.println("生成的CRC-Code为: " + crcStr);

		String remainderStr = GetRemainderStr(sendFrameStr, gxStr);
		
		long remainder = Long.parseLong(remainderStr, 2);
		System.out.println("余数为: " + remainder);
		if (remainder == 0) {
			System.out.println("校验成功");
		} else {
			System.out.println("校验错误");
		}
	}

	public static void main(String[] args) {
		CRC operation = new CRC();
		String frameStr = operation.Send();
		operation.Receive(frameStr);
	}

}

File: 3/3-1/C++/version 2/CRC.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string>
#include <iostream>
using namespace std;

class CRC
{
private:
	string gxStr = "10001000000100001";
public:
	string GetRemainderStr(string dividendStr, string divisorStr)
	{
		int dividendLen = dividendStr.length();
		int divisorLen = divisorStr.length();
		for (int i = 0; i < divisorLen - 1; i++)
		{
			dividendStr += "0";
		}
		for (int i = 0; i < dividendLen; i++)
		{
			if (dividendStr[i] == '1')
			{
				dividendStr[i] = '0';
				for (int j = 1; j < divisorLen; j++)
				{
					if (dividendStr[i + j] == divisorStr[j])
					{
						dividendStr[i + j] = '0';
					}
					else
					{
						dividendStr[i + j] = '1';
					}
				}
			}
		}
		string remainderStr = dividendStr.substr(dividendLen, divisorLen);
		return remainderStr;
	}

	string Send()
	{
		string dataStr = "01100000000000111110000000000110";
		cout << "待发送的数据信息二进制比特串为：" + dataStr << endl;
		cout << "CRC-CCITT对应的二进制比特串为：" + gxStr << endl;

		string remainderStr = GetRemainderStr(dataStr, gxStr);

		string crcStr = remainderStr;
		string sendFrameStr = dataStr + remainderStr;

		cout << "生成的CRC-Code为: " << crcStr << endl;
		cout << "带校验和的发送帧为: " << sendFrameStr << endl << endl;
		return sendFrameStr;
	}

	void Receive(string sendFrameStr)
	{
		int sendFrameLen = sendFrameStr.length();
		int gxLen = gxStr.length();
		string dataStr = sendFrameStr.substr(0, sendFrameLen - gxLen + 1);
		string crcStr = sendFrameStr.substr(sendFrameLen - gxLen + 1);
		cout << "接收的数据信息二进制比特串为：" << dataStr << endl;
		cout << "生成的CRC-Code为: " + crcStr << endl;

		string remainderStr = GetRemainderStr(sendFrameStr, gxStr);

		int remainder = atoi(remainderStr.c_str());
		cout << "余数为: " << remainder << endl;
		if (remainder == 0)
		{
			cout << "校验成功" << endl;
		}
		else
		{
			cout << "校验错误" << endl;
		}
	}

};

int main()
{
	CRC operation;
	string frameStr = operation.Send();
	operation.Receive(frameStr);
	system("pause");
}
